obs-studio-server/source/nodeobs_settings.cpp

```diff
@@ -3984,3 +3990,109 @@ void OBS_settings::saveGenericSettings(std::vector<SubCategory> genericSettings,
 	}
 	config_save_safe(config, "tmp", nullptr);
 }
+
+void getDevices(
+	const char* source_id,
+	const char* property_name,
+	std::vector<ipc::value>& rval)
+{
+	auto settings = obs_get_source_defaults(source_id);
+	if (!settings)
+		return;
+
+	obs_data_set_string(settings, property_name, "dummy_device_id");
+	auto dummy_source = obs_source_create(source_id, "dummy_audio_input", settings, nullptr);
+	if (!dummy_source)
+		return;
+
+	auto props = obs_source_properties(dummy_source);
+	if (!props)
+		return;
+
+	auto prop = obs_properties_get(props, property_name);
+	if (!prop)
+		return;
+
+	size_t items = obs_property_list_item_count(prop);
+	if (rval.size() > 1)
+		rval[1].value_union.ui64 += items;
+	else
+		rval.push_back(ipc::value((uint64_t)items));
+
+	for (size_t idx = 0; idx < items; idx++) {
+		const char* description = obs_property_list_item_name(prop, idx);
+		const char* device_id = obs_property_list_item_string(prop, idx);
+
+		if (!description || !strcmp(description, "") ||
+			!device_id || !strcmp(device_id, "")) {
+			rval[1].value_union.ui64--;
```

Comment:
could this be `rval[0]` in case `rval.size()` is 1 ?

Reply:
Our IPC format requires the first value `rval[0]` to be the error code. That's why I put the size the value after that.